src/main/java/com/redhat/devtools/intellij/lsp4mp4ij/psi/core/java/codeaction/JavaCodeActionContext.java

```diff
@@ -1,138 +1,138 @@
-
-	private final int selectionOffset;
-	private final int selectionLength;
-
-	private final MicroProfileJavaCodeActionParams params;
-	private JavaCodeActionContext source;
-
-	public JavaCodeActionContext(PsiFile typeRoot, int selectionOffset, int selectionLength, IPsiUtils utils,
-								 Module module, MicroProfileJavaCodeActionParams params, JavaCodeActionContext source) {
-		super(params.getUri(), typeRoot, utils, module);
-		this.selectionOffset = selectionOffset;
-		this.selectionLength = selectionLength;
-		this.params = params;
-		this.source = source;
-	}
-
-	public JavaCodeActionContext(PsiFile typeRoot, int selectionOffset, int selectionLength, IPsiUtils utils,
-								 MicroProfileJavaCodeActionParams params) {
-		this(typeRoot, selectionOffset, selectionLength,utils, utils.getModule(), params, null);
-		this.source = this;
-	}
-
-	/**
-	 * Return a copy of the context with its own in memory compilation unit
-	 *
-	 * @return the new context
-	 */
-	public JavaCodeActionContext copy() {
-		return new JavaCodeActionContext(getTypeRoot().getViewProvider().clone().getPsi(getTypeRoot().getLanguage()), selectionOffset,
-				selectionLength, getUtils(), getJavaProject(), params, this.source);
-	}
-
-	public MicroProfileJavaCodeActionParams getParams() {
-		return params;
-	}
-
-	@Override
-	public PsiFile getCompilationUnit() {
-		return getTypeRoot();
-	}
-
-	/**
-	 * Returns the length.
-	 *
-	 * @return int
-	 */
-	@Override
-	public int getSelectionLength() {
-		return selectionLength;
-	}
-
-	/**
-	 * Returns the offset.
-	 *
-	 * @return int
-	 */
-	@Override
-	public int getSelectionOffset() {
-		return selectionOffset;
-	}
-
-	@Override
-	public PsiElement getCoveringNode() {
-		return PsiUtil.getElementInclusiveRange(getASTRoot(), TextRange.from(getSelectionOffset(), getSelectionLength()));
-	}
-
-	@Override
-	public PsiElement getCoveredNode() {
-		return getASTRoot().findElementAt(getSelectionOffset());
-	}
-
-	public JavaCodeActionContext getSource() {
-		return source;
-	}
-
-	public CodeAction convertToCodeAction(ChangeCorrectionProposal proposal, Diagnostic... diagnostics) {
-		String name = proposal.getName();
-		WorkspaceEdit edit = ChangeUtil.convertToWorkspaceEdit(proposal.getChange(), getUri(), getUtils(),
-				params.isResourceOperationSupported());
-		if (!ChangeUtil.hasChanges(edit)) {
-			return null;
-		}
-		ExtendedCodeAction codeAction = new ExtendedCodeAction(name);
-		codeAction.setRelevance(proposal.getRelevance());
-		codeAction.setKind(proposal.getKind());
-		codeAction.setEdit(edit);
-		codeAction.setDiagnostics(Arrays.asList(diagnostics));
-		return codeAction;
-	}
-
-	public WorkspaceEdit convertToWorkspaceEdit(ChangeCorrectionProposal proposal) {
-		WorkspaceEdit edit = ChangeUtil.convertToWorkspaceEdit(proposal.getChange(), getUri(), getUtils(),
-				params.isResourceOperationSupported());
-		if (!ChangeUtil.hasChanges(edit)) {
```

Comment:
ex. a test for `ChangeUtil` could assert the 3 cases present in #hasChanges. Would this make sense?